routing_plugin.go

```diff
@@ -104,3 +120,419 @@ func ReleaseRoutingPlugin(ctx context.Context) {
+
+	queryRoutesReq := lndclient.QueryRoutesRequest{
+		Source:            &r.lnd.NodePubkey,
+		PubKey:            r.target,
+		AmtMsat:           lnwire.MilliSatoshi(r.amount * 1000),
+		FeeLimitMsat:      lnwire.MilliSatoshi(r.amount * 1000),
+		UseMissionControl: true,
+	}
+
+	if log.Level() == btclog.LevelTrace {
+		res, err := r.lnd.Client.QueryRoutes(ctx, queryRoutesReq)
+		log.Tracef("BeforePayment() QueryRoutes(1)=%v, err=%v",
+			spew.Sdump(res), err)
+	}
+
+	// Do not do anything unless we tried to route the payment at least
+	// once.
+	if currAttempt < 2 {
+		return nil
+	}
+
+	// Calculate the limit until we'll disable edges. Anything above this
+	// limit should be encouraged.
+	minFee := r.nodesByMaxFee[0].fee
+	maxFee := r.nodesByMaxFee[len(r.nodesByMaxFee)-1].fee
+	limit := minFee + ((maxFee-minFee)/int64(maxAttempts))*int64(currAttempt)
+
+	// Create a timestamp just slightly in the future as Mission Control
+	// stores timestamps with sub second precision where as we send a unix
+	// timestamp it may occur that we can't override the last entries as
+	// they have the same unix timestamp.
+	// TODO(bhandras): not very reliable, ideally we'd need a force import
+	// for MC.
+	now := r.clock.Now().Add(time.Second)
+
+	allowed := 0
+	var entries []lndclient.MissionControlEntry
+	for _, nodeFeeInfo := range r.nodesByMaxFee {
+		if nodeFeeInfo.fee < limit {
+			log.Debugf("Discouraging payments from %v to %v",
+				nodeFeeInfo.node, r.target)
+			entries = append(
+				entries, lndclient.MissionControlEntry{
+					NodeFrom: nodeFeeInfo.node,
+					NodeTo:   r.target,
+					FailTime: now,
+					FailAmt:  1,
+				})
+		} else {
+			log.Debugf("Encouraging payments from %v to %v",
+				nodeFeeInfo.node, r.target)
+			entries = append(
+				entries, lndclient.MissionControlEntry{
+					NodeFrom:    nodeFeeInfo.node,
+					NodeTo:      r.target,
+					SuccessTime: now,
+					SuccessAmt: lnwire.MilliSatoshi(
+						nodeFeeInfo.capacity * 1000,
+					),
+				})
+			allowed++
+		}
+	}
+
+	// There's no point retrying the payment since we discouraged using
+	// all inbound peers to the target.
+	if allowed == 0 {
+		return ErrRoutingPluginNoMoreRetries
+	}
+
+	err := r.lnd.Router.ImportMissionControl(ctx, entries)
+	log.Tracef("Imported MC state: %v, err=%v",
+		spew.Sdump(entries), err)
+
+	if err == nil && log.Level() == btclog.LevelTrace {
+		res, err := r.lnd.Client.QueryRoutes(ctx, queryRoutesReq)
+		log.Tracef("BeforePayment() QueryRoutes(2)=%v, err=%v",
+			spew.Sdump(res), err)
+	}
+
+	return err
+}
+
+// Done will reset the mission control state to where it was before using the
+// routing plugin.
+func (r *lowToHighRoutingPlugin) Done(ctx context.Context) error {
+	if r.mcState == nil {
+		return nil
+	}
+
+	var err error
+	if len(r.mcState) > 0 {
+		now := r.clock.Now()
+		// Roll the entry times forward (to be able to override recent
+		// updates). TODO: maybe we should reset here first instead.
```

Comment:
(re TODO): I think that resetting first might be a bit risky, because an inconvenient restart could completely wipe MC (and we lose data).

Reply:
Yes good point.